lib/data_table.rb

```diff
@@ -977,6 +943,58 @@ def parameter_binding_character
         is_connection_sql_server? ? /[\$@].+/ : /\$.+/
       end
 
+      def adjust_user_visibility(columns, options)
+        return unless columns.present? && options.dig(:columns_layout, 'columns')
+
+        columns.each do |column|
+          user_column = find_user_column(column.name, options)
+          column.options.reverse_merge!(:user_visible => user_column['visible'], :user_visible_index => user_column['visibleIndex'])
+        end
+
+        columns.sort_by! { |c| c.options[:user_visible_index] }
+      end
+
+      def find_user_column(column_name, options)
+        options[:columns_layout]['columns'].detect do |c|
+          split_layout_name = c['dataField'].split('.')
+          name = split_layout_name.length > 2 ? split_layout_name.last(2).map(&:to_sym).to_s : split_layout_name.last.to_s
+          column_name.to_s == name
+        end || { 'visible' => false }
+      end
+
+      def generate_csv_output_stream(csv_output_stream, columns)
+        if @options.fetch(:write_headers, true)
+          csv_output_stream << columns.map(&:caption)
+        end
+
+        connection = ActiveRecord::Base.connection.raw_connection
+        results = connection.exec(@base_query.to_sql)
+
+        results.each do |result|
+          csv_output_stream << columns.map { |column| result[column.name.to_s] }
+        end
+      end
+
+      def generate_csv_content(view_context, params, options, columns)
+        header = []
+        rows = []
+
+        if @options.fetch(:write_headers, true)
+          header << columns.map(&:caption).join(',')
+        end
+
+        resultset, _ = get_resultset_and_count(params, options)
+
+        resultset.each do |instance|
```

Comment:
This materializes the view. Need to use something like:

`copy_connection.copy_data "/*NO LOAD BALANCE*/ COPY (#{sql}) TO STDOUT WITH (#{options_string})", decoder do
      while line = copy_connection.get_copy_data
        # replacing non-UTF-8 chars with '??'
        temp_file.write line.join(delimiter).encode(encoding, :replace => '??', :invalid => :replace, :undef => :replace)
        temp_file.write "\n"
      end
    end`

Reply:
(found in csv_write.rb)